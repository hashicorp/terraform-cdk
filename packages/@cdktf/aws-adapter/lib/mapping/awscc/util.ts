import { toCamelCase } from "codemaker";
// TODO: expose the name conflict replacement functionality that is used from this in a better way
import { AttributeModel } from "cdktf-cli/lib/get/generator/models/attribute-model";

const propertyNameReplacements = {
  CloudFormation: "Cloudformation",
  CloudFront: "Cloudfront",
  CloudWatch: "Cloudwatch",
  FSx: "Fsx",
};

export function convertCloudFormationPropertyToCDKTFAttribute(
  cfnAttributeName: string
): string {
  const tfAttribute =
    convertCloudFormationPropertyToTFAttribute(cfnAttributeName);

  const cdktfAttribute = toCamelCase(tfAttribute);

  const tmp = new AttributeModel({
    name: cdktfAttribute,
    computed: false,
    optional: false,
    provider: false,
    required: false,
    storageName: "",
    terraformFullName: "",
    terraformName: "",
    type: null as any,
  });

  // tmp.name is getter and does more than returning the name we supplied as input
  const conflictFreeCdktfAttribute = tmp.name;

  return conflictFreeCdktfAttribute;
}

// reimplementation of naming.go from terraform-provider-awscc
// TODO: use WebAssembly to reuse naming.go
export function convertCloudFormationPropertyToTFAttribute(
  cfnAttributeName: string
): string {
  let propertyName = cfnAttributeName.trim();

  if (propertyName === "") {
    return propertyName;
  }

  for (let replacement in Object.entries(propertyNameReplacements)) {
    const [old, new_] = replacement;
    propertyName = propertyName.replace(new RegExp(old, "g"), new_);
  }

  let attributeName = "";

  for (let i = 0; i < propertyName.length; i++) {
    let ch = propertyName.charAt(i);

    const isCap = isCapitalLetter(ch);
    const isLow = isLowercaseLetter(ch);
    const isDig = isNumeric(ch);

    if (isCap) {
      ch = toLowercaseLetter(ch);
    }

    if (i < propertyName.length - 1) {
      let nextCh = propertyName[i + 1];
      let nextIsCap = isCapitalLetter(nextCh);
      let nextIsLow = isLowercaseLetter(nextCh);
      let nextIsDig = isNumeric(nextCh);

      // Append underscore if case changes.
      if (
        (isCap && nextIsLow) ||
        (isLow && (nextIsCap || nextIsDig)) ||
        (isDig && (nextIsCap || nextIsLow))
      ) {
        if (isCap && nextIsLow) {
          const prevIsCap = i > 0 && isCapitalLetter(propertyName[i - 1]);
          if (prevIsCap) {
            attributeName += "_";
          }
        }
        attributeName += ch;
        if (isLow || isDig) {
          attributeName += "_";
        }

        continue;
      }
    }

    if (isCap || isLow || isDig) {
      attributeName += ch;
    } else {
      attributeName += "_";
    }
  }

  return attributeName;
}

function isCapitalLetter(ch: string): boolean {
  return ch >= "A" && ch <= "Z";
}
function isLowercaseLetter(ch: string): boolean {
  return ch >= "a" && ch <= "z";
}
function isNumeric(ch: string): boolean {
  return ch >= "0" && ch <= "9";
}
function toLowercaseLetter(ch: string): string {
  return ch.toLowerCase();
}
