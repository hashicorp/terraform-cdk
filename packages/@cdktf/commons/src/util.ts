// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { SpawnOptions } from "child_process";
import { spawn } from "cross-spawn";
import * as fs from "fs-extra";
import { https, http } from "follow-redirects";
import * as os from "os";
import * as path from "path";
import { processLoggerError, processLoggerDebug } from "./logging";
import { IManifest, Manifest } from "cdktf";
import * as config from "./config";

export async function shell(
  program: string,
  args: string[] = [],
  options: SpawnOptions = {}
) {
  const stderr = new Array<string | Uint8Array>();
  const stdout = new Array<string>();
  try {
    return await exec(
      program,
      args,
      options,
      (chunk: Buffer) => {
        stdout.push(chunk.toString());
        console.log(chunk.toString());
      },
      (chunk: string | Uint8Array) => stderr.push(chunk)
    );
  } catch (e: any) {
    if (stderr.length > 0) {
      e.stderr = stderr.map((chunk) => chunk.toString()).join("");
    }
    if (stdout.length > 0) {
      e.stdout = stdout.join("");
    }
    throw e;
  }
}

export async function withTempDir(
  dirname: string,
  closure: () => Promise<void>
) {
  const prevdir = process.cwd();
  const parent = await fs.mkdtemp(path.join(os.tmpdir(), "cdktf."));
  const workdir = path.join(parent, dirname);
  await fs.mkdirp(workdir);
  try {
    process.chdir(workdir);
    await closure();
  } finally {
    process.chdir(prevdir);
    await fs.remove(parent);
  }
}

export async function mkdtemp(closure: (dir: string) => Promise<void>) {
  const workdir = await fs.mkdtemp(path.join(os.tmpdir(), "cdktf."));
  try {
    await closure(workdir);
  } finally {
    await fs.remove(workdir);
  }
}

export const exec = async (
  command: string,
  args: string[],
  options: SpawnOptions,
  stdout?: (chunk: Buffer) => any,
  stderr?: (chunk: string | Uint8Array) => any,
  sendToStderr = true
): Promise<string> => {
  return new Promise((ok, ko) => {
    const child = spawn(command, args, options);
    const out = new Array<Buffer>();
    const err = new Array<string | Uint8Array>();
    if (stdout !== undefined) {
      child.stdout?.on("data", (chunk: Buffer) => {
        processLoggerDebug(chunk);
        out.push(chunk);
        stdout(chunk);
      });
    } else {
      child.stdout?.on("data", (chunk: Buffer) => {
        processLoggerDebug(chunk);
        out.push(chunk);
      });
    }
    if (stderr !== undefined) {
      child.stderr?.on("data", (chunk: string | Uint8Array) => {
        processLoggerError(chunk);
        if (sendToStderr) {
          stderr(chunk);
        }
        err.push(chunk);
      });
    } else {
      child.stderr?.on("data", (chunk: string | Uint8Array) => {
        processLoggerError(chunk);
        if (sendToStderr) {
          process.stderr.write(chunk);
        }
        err.push(chunk);
      });
    }
    child.once("error", (err: any) => ko(err));
    child.once("close", (code: number) => {
      if (code !== 0) {
        const error = new Error(`non-zero exit code ${code}`);
        (error as any).stderr = err.map((chunk) => chunk.toString()).join("");
        return ko(error);
      }
      return ok(Buffer.concat(out).toString("utf-8"));
    });
  });
};

export async function readCDKTFVersion(outputDir: string): Promise<string> {
  const outputFile = path.join(outputDir, "cdk.tf.json");
  if (fs.existsSync(outputFile)) {
    const outputJSON = fs.readFileSync(outputFile, "utf8");
    const data = JSON.parse(outputJSON);
    return data["//"].metadata.version;
  }

  return "";
}

export async function readCDKTFManifest(): Promise<IManifest> {
  const { output } = config.readConfigSync();
  const json = await fs.readFile(path.join(output, Manifest.fileName));
  return JSON.parse(json.toString()) as IManifest;
}

/**
 * Downcase the first character in a string.
 *
 * @param str the string to be processed.
 */
export function downcaseFirst(str: string): string {
  if (str === "") {
    return str;
  }
  return `${str[0].toLocaleLowerCase()}${str.slice(1)}`;
}

export class HttpError extends Error {
  constructor(message?: string, public statusCode?: number) {
    super(message); // 'Error' breaks prototype chain here
    Object.setPrototypeOf(this, new.target.prototype); // restore prototype chain
    // see: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#support-for-newtarget
  }
}

export async function downloadFile(
  url: string,
  targetFilename: string
): Promise<void> {
  // if the type is inferred to be "http|https" calling .get() is not possible
  // because the options parameter (which we don't use anyway) for get is
  // not compatible between http and https -> so we treat it as http
  const client = (url.startsWith("http://") ? http : https) as typeof http;
  const file = fs.createWriteStream(targetFilename);
  return new Promise((ok, ko) => {
    const request = client.get(url, (response) => {
      if (response.statusCode !== 200) {
        ko(
          new HttpError(
            `Failed to get '${url}' (${response.statusCode})`,
            response.statusCode
          )
        );
        return;
      }
      response.pipe(file);
    });

    file.on("finish", () => ok());

    request.on("error", (err: Error) => {
      fs.unlink(targetFilename, () => ko(err));
    });

    file.on("error", (err) => {
      fs.unlink(targetFilename, () => ko(err));
    });

    request.end();
  });
}

/**
 * Awaits a promise and makes sure it's error (if any) is only thrown after all other promises are settled
 * if the promise does not throw an error, the other promises won't be awaited
 * @param p promise to await
 * @param promises promises to await to be all settled if p failed before throwing error that p failed with
 */
export async function ensureAllSettledBeforeThrowing(
  p: Promise<any>,
  promises: (Promise<any> | undefined)[]
) {
  try {
    await p;
  } catch (e) {
    // if an error happened, we still need to wait for all other promises that
    // are currently in progress to complete to allow them to properly wrap up
    await Promise.allSettled(promises);
    throw e;
  }
}
