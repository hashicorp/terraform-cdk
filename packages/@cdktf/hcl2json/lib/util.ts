/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

/**
 * This function is used to escape quotes within a string.
 *  Note: This is a really naive implmentation that will not be able to work with
 *    nested templates. However, nested templates require parsing, so we will
 *    only handle strings without nested templates for the moment
 * @param input The string to escape
 */
export function replaceQuotes(input: string): string {
  const fullyQuotedMatch = /^(\s*")(.*)("\s*)$/.exec(input);
  const hasTemplate = /(?<!\$)\$\{/.test(input);

  if (fullyQuotedMatch !== null && !hasTemplate) {
    return `${fullyQuotedMatch[1]}${fullyQuotedMatch[2].replace(/"/g, `\\"`)}${
      fullyQuotedMatch[3]
    }`;
  }

  return input;
}

/**
 * This function is used to wrap a string in quotes conditionally.
 *  - If the string is already quoted, it will be returned as-is.
 *  - If the string contains a newline, it will be wrapped in a heredoc.
 *  - If the string is a number, it will be returned as-is.
 *  - If the string is a boolean, it will be returned as-is.
 *  - If the string is a list or map, it will be returned as-is.
 *
 * @param input The string to wrap
 * @returns The wrapped string, and the offset of the cursor within the string
 *
 * @example
 * wrapTerraformExpression("foo") // { wrap: '"foo"', wrapOffset: 1 }
 * wrapTerraformExpression("foo\nbar") // { wrap: '<<EOE\nfoo\nbar\nEOE\n', wrapOffset: 5 }
 * wrapTerraformExpression("22") // { wrap: '22', wrapOffset: 0 }
 * wrapTerraformExpression("true") // { wrap: 'true', wrapOffset: 0 }
 * wrapTerraformExpression("[22, 10]") // { wrap: '[22, 10]', wrapOffset: 0 }
 * wrapTerraformExpression("{ foo = 22 }") // { wrap: '{ foo = 22 }', wrapOffset: 0 }
 */
export function wrapTerraformExpression(input: string): {
  wrap: string;
  wrapOffset: number;
} {
  input = replaceQuotes(input);

  if (!isNaN(Number(input))) {
    return { wrap: input, wrapOffset: 0 };
  }
  if (input === "true" || input === "false") {
    return { wrap: input, wrapOffset: 0 };
  }

  if (input.startsWith(`"`)) {
    if (input.indexOf("\n") >= 0) {
      const trimWrapped = input.substring(1, input.length - 1);
      return {
        wrap: `<<EOE\n${trimWrapped}\nEOE\n`,
        wrapOffset: 5,
      };
    }
    return { wrap: input, wrapOffset: 0 };
  }

  if (input.startsWith("[") || input.startsWith("{")) {
    return { wrap: input, wrapOffset: 0 };
  }

  if (input.startsWith("<<")) {
    // For Heredocs, we need to ensure that the string ends with an empty newline as
    // the HCL parser doesn't find the ending token otherwise
    if (!input.endsWith("\n")) {
      return { wrap: input + "\n", wrapOffset: 0 };
    }

    return { wrap: input, wrapOffset: 0 };
  }

  return { wrap: `"${input}"`, wrapOffset: 1 };
}
