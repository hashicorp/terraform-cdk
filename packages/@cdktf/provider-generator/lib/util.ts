// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { spawn, SpawnOptions } from "child_process";
import * as fs from "fs-extra";
import * as os from "os";
import * as path from "path";

export async function withTempDir(
  dirname: string,
  closure: () => Promise<void>
) {
  const prevdir = process.cwd();
  const parent = await fs.mkdtemp(path.join(os.tmpdir(), "cdktf."));
  const workdir = path.join(parent, dirname);
  await fs.mkdirp(workdir);
  try {
    process.chdir(workdir);
    await closure();
  } finally {
    process.chdir(prevdir);
    await fs.remove(parent);
  }
}

export async function mkdtemp(closure: (dir: string) => Promise<void>) {
  const workdir = await fs.mkdtemp(path.join(os.tmpdir(), "cdktf."));
  try {
    await closure(workdir);
  } finally {
    await fs.remove(workdir);
  }
}

export const exec = async (
  command: string,
  args: string[],
  options: SpawnOptions,
  stdout?: (chunk: Buffer) => any,
  stderr?: (chunk: string | Uint8Array) => any
): Promise<string> => {
  return new Promise((ok, ko) => {
    const child = spawn(command, args, options);
    const out = new Array<Buffer>();
    const err = new Array<string | Uint8Array>();
    if (stdout !== undefined) {
      child.stdout?.on("data", (chunk: Buffer) => {
        stdout(chunk);
      });
    } else {
      child.stdout?.on("data", (chunk: Buffer) => {
        out.push(chunk);
      });
    }
    if (stderr !== undefined) {
      child.stderr?.on("data", (chunk: string | Uint8Array) => {
        stderr(chunk);
      });
    } else {
      child.stderr?.on("data", (chunk: string | Uint8Array) => {
        process.stderr.write(chunk);
        err.push(chunk);
      });
    }
    child.once("error", (err: any) => ko(err));
    child.once("close", (code: number) => {
      if (code !== 0) {
        const error = new Error(`non-zero exit code ${code}`);
        (error as any).stderr = err.map((chunk) => chunk.toString()).join("");
        return ko(error);
      }
      return ok(Buffer.concat(out).toString("utf-8"));
    });
  });
};

/**
 * Downcase the first character in a string.
 *
 * @param str the string to be processed.
 */
export function downcaseFirst(str: string): string {
  if (str === "") {
    return str;
  }
  return `${str[0].toLocaleLowerCase()}${str.slice(1)}`;
}

/**
 * Uppercase the first character in a string.
 *
 * @param str the string to be processed.
 */
export function uppercaseFirst(str: string): string {
  if (str === "") {
    return str;
  }
  return `${str[0].toLocaleUpperCase()}${str.slice(1)}`;
}
