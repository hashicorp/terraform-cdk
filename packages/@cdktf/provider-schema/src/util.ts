/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import { spawn, SpawnOptions } from "child_process";
import * as fs from "fs-extra";
import * as os from "os";
import * as path from "path";

export async function withTempDir(
  dirname: string,
  closure: () => Promise<void>
) {
  const prevdir = process.cwd();
  const parent = await fs.mkdtemp(path.join(os.tmpdir(), "cdktf."));
  const workdir = path.join(parent, dirname);
  await fs.mkdirp(workdir);
  try {
    process.chdir(workdir);
    await closure();
  } finally {
    process.chdir(prevdir);
    await fs.remove(parent);
  }
}

export const exec = async (
  command: string,
  args: string[],
  options: SpawnOptions,
  stdout?: (chunk: Buffer) => any,
  stderr?: (chunk: string | Uint8Array) => any
): Promise<string> => {
  return new Promise((ok, ko) => {
    const child = spawn(command, args, options);
    const out = new Array<Buffer>();
    const err = new Array<string | Uint8Array>();
    if (stdout !== undefined) {
      child.stdout?.on("data", (chunk: Buffer) => {
        stdout(chunk);
      });
    } else {
      child.stdout?.on("data", (chunk: Buffer) => {
        out.push(chunk);
      });
    }
    if (stderr !== undefined) {
      child.stderr?.on("data", (chunk: string | Uint8Array) => {
        stderr(chunk);
      });
    } else {
      child.stderr?.on("data", (chunk: string | Uint8Array) => {
        process.stderr.write(chunk);
        err.push(chunk);
      });
    }
    child.once("error", (err: any) => ko(err));
    child.once("close", (code: number) => {
      if (code !== 0) {
        const error = new Error(`non-zero exit code ${code}`);
        (error as any).stderr = err.map((chunk) => chunk.toString()).join("");
        return ko(error);
      }
      return ok(Buffer.concat(out).toString("utf-8"));
    });
  });
};
