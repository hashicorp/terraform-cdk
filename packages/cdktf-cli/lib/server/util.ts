import parseGitignore from "parse-gitignore";
import * as path from "path";
import { promises as fs } from "fs";
import { WatchErrorOrigin, WatchState } from "./WatchClient";
import {
  DeployingResourceApplyState,
  PlannedResourceAction,
} from "../models/terraform";

export async function readGitignore(directory: string): Promise<string[]> {
  const content = await fs.readFile(path.join(directory, ".gitignore"));
  return parseGitignore(content);
}

// todo: generate this via some graphql tool
export type GraphQLWatchState = {
  status: WatchState["status"];
  stacks: Array<{
    name: string;
    constructPath: string;
    synthesizedStackPath: string;
    workingDirectory: string;
  }>;
  resources: Array<{
    id: string;
    action: "CREATE" | "UPDATE" | "DELETE" | "READ" | "NO_OP";
    deployState:
      | "WAITING"
      | "UPDATING"
      | "UPDATED"
      | "CREATING"
      | "CREATED"
      | "DESTROYING"
      | "DESTROYED"
      | "SUCCESS"
      | "ERROR";
    changedAt: number;
  }>;
  error?: {
    message: string;
    recoverable: boolean;
    origin: WatchErrorOrigin | "SERVER";
    timestamp: number;
  };
};

export function mapWatchState(state: WatchState): GraphQLWatchState {
  return {
    status: state.status,
    stacks: state.stacks.map((s) => ({
      constructPath: s.constructPath,
      name: s.name,
      synthesizedStackPath: s.synthesizedStackPath,
      workingDirectory: s.workingDirectory,
    })),
    resources: Array.from(state.resources.values()).map((r) => ({
      id: r.id,
      action: plannedResourceActionReverseMap[r.action],
      deployState: deployingResourceApplyStateReverseMap[r.applyState],
      changedAt: r.changedAt,
    })),
    error: state.error
      ? {
          message: state.error.message,
          origin: state.error.origin,
          recoverable: state.error.recoverable,
          timestamp: state.error.timestamp,
        }
      : undefined,
  };
}

const plannedResourceActionReverseMap: Record<
  PlannedResourceAction,
  GraphQLWatchState["resources"][number]["action"]
> = {
  "no-op": "NO_OP",
  create: "CREATE",
  delete: "DELETE",
  read: "READ",
  update: "UPDATE",
};

const deployingResourceApplyStateReverseMap: Record<
  DeployingResourceApplyState,
  GraphQLWatchState["resources"][number]["deployState"]
> = {
  waiting: "WAITING",
  updating: "UPDATING",
  updated: "UPDATED",
  creating: "CREATING",
  created: "CREATED",
  destroying: "DESTROYING",
  destroyed: "DESTROYED",
  success: "SUCCESS",
  error: "ERROR",
};
