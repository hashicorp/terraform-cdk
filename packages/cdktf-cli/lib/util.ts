import * as http from 'http';
import * as https from 'https';
import { spawn, SpawnOptions } from 'child_process';
import * as fs from 'fs-extra';
import * as os from 'os';
import * as path from 'path';
import { processLogger } from './logging';

export async function shell(program: string, args: string[] = [], options: SpawnOptions = { }) {
  return new Promise((ok, ko) => {
    const child = spawn(program, args, { stdio: 'inherit', ...options });
    child.once('error', ko);
    child.once('exit', code => {
      if (code === 0) { return ok(); }
      else { return ko(new Error(`non-zero exit code ${code}`)); }
    });
  });
}

export async function withTempDir(dirname: string, closure: () => Promise<void>) {
  const prevdir = process.cwd();
  const parent = await fs.mkdtemp(path.join(os.tmpdir(), 'cdktf.'));
  const workdir = path.join(parent, dirname);
  await fs.mkdirp(workdir);
  try {
    process.chdir(workdir);
    await closure();
  } finally {
    process.chdir(prevdir);
    await fs.remove(parent);
  }
}

async function get(url: string, protocol: typeof http | typeof https = https): Promise<string> {
  return new Promise((ok, ko) => {
    const req = protocol.get(url, res => {
      if (res.statusCode !== 200) {
        throw new Error(`${res.statusMessage}: ${url}`);
      }
      const data = new Array<Buffer>();
      res.on('data', chunk => data.push(chunk));
      res.once('end', () => ok(Buffer.concat(data).toString('utf-8')));
      res.once('error', ko);
    });

    req.once('error', ko);
    req.end();
  });
}

export async function httpGet(url: string): Promise<string> {
  return get(url, http)
}

export async function httpsGet(url: string): Promise<string> {
  return get(url)
}

export const exec = async (command: string, args: string[], options: SpawnOptions, stdout?: (chunk: Buffer) => any): Promise<string> => {
  return new Promise((ok, ko) => {
    const child = spawn(command, args, options);
    const out = new Array<Buffer>();
    if (stdout !== undefined) {
      child.stdout?.on('data', (chunk: Buffer) => { processLogger(chunk) ; stdout(chunk) });
    } else {
      child.stdout?.on('data', (chunk: Buffer) =>  { processLogger(chunk) ; out.push(chunk) } );
    }
    child.stderr?.on('data', (chunk: string | Uint8Array) => { processLogger(chunk) ; process.stderr.write(chunk) });
    child.once('error', (err: any) => ko(err));
    child.once('close', (code: number) => {
      if (code !== 0) {
        return ko(new Error(`non-zero exit code ${code}`));
      }
      return ok(Buffer.concat(out).toString('utf-8'));
    });
  });
}
