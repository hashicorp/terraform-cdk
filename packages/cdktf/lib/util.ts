// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import {
  sourceOrTargetNotAnObject,
  targetNotResolvableWithOverrides,
} from "./errors";
import { TerraformDynamicBlock } from "./terraform-dynamic-block";
import { Tokenization } from "./tokens/token";

export const terraformBinaryName =
  process.env.TERRAFORM_BINARY_NAME || "terraform";

/**
 * Merges `source` into `target`, overriding any existing values.
 * `undefined` will cause a value to be deleted.
 */
export function deepMerge(target: any, ...sources: any[]) {
  if (Tokenization.isResolvable(target) && sources.length > 0) {
    throw targetNotResolvableWithOverrides(target.toString());
  }

  for (const source of sources) {
    if (typeof source !== "object" || typeof target !== "object") {
      throw sourceOrTargetNotAnObject(
        JSON.stringify(source),
        typeof source,
        JSON.stringify(target),
        typeof target
      );
    }

    for (const key of Object.keys(source)) {
      const value = source[key];
      if (typeof value === "object" && value != null && !Array.isArray(value)) {
        // if the value at the target is not an object, override it with an
        // object so we can continue the recursion
        if (typeof target[key] !== "object") {
          target[key] = {};
        }

        // if the value is a resolvable we don't want to recurse into it
        if (Tokenization.isResolvable(value)) {
          target[key] = value;
          continue;
        }

        deepMerge(target[key], value);
      } else if (
        typeof value === "object" &&
        value != null &&
        Array.isArray(value)
      ) {
        if (Array.isArray(target[key])) {
          target[key] = [...target[key], ...value];
        } else {
          target[key] = value;
        }
      } else if (value === undefined) {
        delete target[key];
      } else {
        target[key] = value;
      }
    }
  }

  return target;
}

/**
 * Transforms a string to snake case
 */
export function snakeCase(str: string): string {
  if (!str) return "";

  return String(str)
    .replace(/^[^A-Za-z0-9]*|[^A-Za-z0-9]*$/g, "")
    .replace(/([a-z0-9])([A-Z])/g, (_m, a, b) => a + "_" + b.toLowerCase())
    .replace(/[^A-Za-z0-9]+|_+/g, "_")
    .toLowerCase();
}

/**
 * Transforms all keys in a object to snake case
 */
export function keysToSnakeCase(object: any): any {
  if (Tokenization.isResolvable(object)) {
    return object;
  }
  if (Array.isArray(object)) {
    return object.map((e: any) => {
      return typeof e === "object" ? keysToSnakeCase(e) : e;
    });
  }
  const keys = Object.keys(object);
  return keys.reduce((newObject: any, key: string) => {
    if (key === "tags" || key === "environment") {
      newObject[key] = object[key];
      return newObject;
    }

    let value = object[key];
    if (typeof value === "object") {
      value = keysToSnakeCase(value);
    }
    if (Array.isArray(value)) {
      value = value.map((e: any) => {
        return typeof e === "object" ? keysToSnakeCase(e) : e;
      });
    }
    newObject[snakeCase(key)] = value;
    return newObject;
  }, {});
}

/**
 * dynamic attributes are located at a different position than normal block attributes
 * This method detects them and moves them from .attributeName to .dynamic.attributeName
 * It also invokes the .toTerraform() method on the dynamic attribute to get the correct
 * Terraform representation
 */
export function processDynamicAttributesForHcl(attributes: {
  [name: string]: any;
}): {
  [name: string]: any;
} {
  const result: { [name: string]: any; dynamic?: { [name: string]: any } } = {};
  Object.entries(attributes).forEach(([attributeName, value]) => {
    if (TerraformDynamicBlock.isTerraformDynamicBlock(value)) {
      if (!result.dynamic) {
        result.dynamic = {};
      }
      result.dynamic[attributeName] = value.toTerraformDynamicBlockJson();
    } else {
      const recurse =
        value !== null &&
        typeof value === "object" &&
        value.constructor === Object; // only descend into plain objects
      result[attributeName] = recurse
        ? processDynamicAttributesForHcl(value)
        : value;
    }
  });
  return result;
}

/**
 * dynamic attributes are located at a different position than normal block attributes
 * This method detects them and moves them from .attributeName to .dynamic.attributeName
 * It also invokes the .toTerraform() method on the dynamic attribute to get the correct
 * Terraform representation
 */
export function processDynamicAttributes(attributes: { [name: string]: any }): {
  [name: string]: any;
} {
  const result: { [name: string]: any; dynamic?: { [name: string]: any } } = {};
  Object.entries(attributes).forEach(([attributeName, value]) => {
    if (TerraformDynamicBlock.isTerraformDynamicBlock(value)) {
      if (!result.dynamic) {
        result.dynamic = {};
      }
      result.dynamic[attributeName] = value.toTerraformDynamicBlockJson();
    } else {
      const recurse =
        value !== null &&
        typeof value === "object" &&
        value.constructor === Object; // only descend into plain objects
      result[attributeName] = recurse ? processDynamicAttributes(value) : value;
    }
  });
  return result;
}
