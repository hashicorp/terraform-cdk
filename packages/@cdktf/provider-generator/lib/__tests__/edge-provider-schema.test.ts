// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import * as fs from "fs-extra";
import * as path from "path";
import { generateProviderBindingsFromSchema } from "..";
import { mkdtemp } from "../util";
import { edgeSchema } from "./edge-provider-schema";

describe("Edge Provider Schema", () => {
  it("compiles to Typescript", async () => {
    return mkdtemp(async (dir) => {
      await generateProviderBindingsFromSchema(dir, edgeSchema);

      fs.readdirSync(path.join(dir, "providers", "edge")).forEach((file) => {
        if (!file.endsWith(".ts")) {
          return;
        }
        expect(
          fs.readFileSync(path.join(dir, "providers", "edge", file), "utf8")
        ).toMatchSnapshot(file);
      });
    });
  });
});
