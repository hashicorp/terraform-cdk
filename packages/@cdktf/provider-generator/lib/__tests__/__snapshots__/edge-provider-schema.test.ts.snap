// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edge Provider Schema compiles to Typescript: index.ts 1`] = `
"// generated by cdktf get
export * as requiredAttributeResource from './required-attribute-resource';
export * as optionalAttributeResource from './optional-attribute-resource';
export * as optionalComputedAttributeResource from './optional-computed-attribute-resource';
export * as listBlockResource from './list-block-resource';
export * as mapResource from './map-resource';
export * as setBlockResource from './set-block-resource';
export * as mapListResource from './map-list-resource';
export * as provider from './provider';

"
`;

exports[`Edge Provider Schema compiles to Typescript: lazy-index.ts 1`] = `
"// generated by cdktf get
Object.defineProperty(exports, 'requiredAttributeResource', { get: function () { return require('./required-attribute-resource'); } });
Object.defineProperty(exports, 'optionalAttributeResource', { get: function () { return require('./optional-attribute-resource'); } });
Object.defineProperty(exports, 'optionalComputedAttributeResource', { get: function () { return require('./optional-computed-attribute-resource'); } });
Object.defineProperty(exports, 'listBlockResource', { get: function () { return require('./list-block-resource'); } });
Object.defineProperty(exports, 'mapResource', { get: function () { return require('./map-resource'); } });
Object.defineProperty(exports, 'setBlockResource', { get: function () { return require('./set-block-resource'); } });
Object.defineProperty(exports, 'mapListResource', { get: function () { return require('./map-list-resource'); } });
Object.defineProperty(exports, 'provider', { get: function () { return require('./provider'); } });

"
`;
